argcomplete/bash_completion.d/python-argcomplete.sh

```diff
@@ -34,21 +47,21 @@ _python_argcomplete_global() {
             else
                 return
             fi
-        elif [[ -f "${COMP_WORDS[1]}" ]] && (head -c 1024 "${COMP_WORDS[1]}" | grep --quiet "PYTHON_ARGCOMPLETE_OK") >/dev/null 2>&1; then
+        elif [[ -f "${COMP_WORDS[1]}" ]] && __python_argcomplete_scan_head_noerr "${COMP_WORDS[1]}"; then
             local ARGCOMPLETE=2
         else
             return
         fi
-    elif which "$executable" >/dev/null 2>&1; then
-        local SCRIPT_NAME=$(which "$executable")
+    elif type -P "$executable" >/dev/null 2>&1; then
+        local SCRIPT_NAME=$(type -P "$executable")
```

Comment:
I don't think this is quite right now - `type -P` will find a file even if shadowed by something else, causing the below block to search that file for the marker, but if it finds it, it will not actually be able to run it. Unfortunately `type -p "$executable"` appears to exit with code 0 if the thing exists but is not a file, so you'd have to check if its output is not blank, or alternatively check if `type -t "$executable"` is "file".

Reply:
I'm not sure if I follow. Can you give a specific situation where `type -P` behaves differently from `which`? Because to me it seems that `which` is just as susceptible to the problem you describe.

Reply:
My mistake - you're right that they both suffer from the same problem. In which case what I'm suggesting is actually an improvement over what we had before, since `type -p` is able to tell whether the thing you're trying to execute is actually a file.

As an example, suppose you create an alias which shadows a script on your path that is able to be completed by argcomplete. If the user types `myalias <tab>`, completion will complete the script, however when the user actually tries to execute it, it will execute the alias.

(As an aside, I've just noticed that currently we're wrapping `$executable` in quotes when we execute it, which is actually incorrect, and means we won't correctly operate on things like aliases and things with quotes and escaped characters. If we remove that, we'll actually execute the alias during tab completion, which will be even worse.)

Reply:
I dug into this a little further and realised it's not really to do with whether `$executable` is quoted or not. We'd need to do something like [this](https://stackoverflow.com/questions/6988288/linux-bash-how-to-unquote) if we want aliases and special characters to work.